index.ts

```diff
@@ -51,35 +85,184 @@ const checkUser = (checkedUser: User, users: Object) => {
+    if (arr.length === 0) {
+        return 0;
+    }
+    // for (let i = 1; i <= Math.max(...arr) + 1; i++) {
+    //     // console.log(i + "|" + arr.indexOf(+i) + "|" + Math.max(...arr));
+    //     if (arr.indexOf(+i) === -1) {
+    //         // console.log("done");
+    //         return +i;
+    //     }
+    // }
+    return Math.max(...arr) + 1;
+};
+
 const sendToClients = (category, data) => {
     wss.clients.forEach(function each(ws) {
         ws.send(JSON.stringify([category, data]));
     })
 };
 
+const updateMessagesFromDb = async () => {
+    console.log("Getting messages from database...");
+    const rows = await knex.from("messages").select("*");
+    let messageIds: Array<number> = [];
+    for (let message of rows) {
+        let userFound = false;
+        let msgUser: any;
+        let takenIds: Array<number> = [];
+        const res = await knex("users").where({id: message.userid});
+        for (let user of res) {
+            takenIds.push(user.id);
+            if (user.id === message.userid) {
+                userFound = true;
+                msgUser = user;
+            }
+        }
+        highestUserId = getHighestFromArr(takenIds);
+        let messageUser: User;
+        if (userFound === true) {
+            messageUser = {
+                username: msgUser.username,
+                userNum: msgUser.usernum,
+                currentChannel: msgUser.currentChannel,
+                id: msgUser.id,
+                messages: []
+            };
+        } else {
+            messageUser = {
+                username: "DeletedUser",
+                userNum: 9999,
+                currentChannel: 0,
+                id: 0,
+                messages: [],
+            };
+        }
+        const newMessage: Message = {
+            user: messageUser,
+            utctime: message.date,
+            message: message.message,
+            channel: message.channel,
+            id: message.id
+        };
+        messages.push(newMessage);
+        messageIds.push(message.id);
+    }
+    highestId = getHighestFromArr(messageIds);
+};
+
+const updateChannelsFromDb = async () => {
+    console.log("Getting channels from database...");
+    try {
+        const rows = await knex.from("channels").select("*");
+        for (const channel of rows) {
+            // console.log(channel);
+            channels[channel.id] = {
+                name: channel.name,
+                id: channel.id,
+                messages: channel.messages
+            };
+        }
+    } catch (err) {
+        console.error(err);
+        throw err;
+    }
+};
+
+const deleteOldUsers = (sNum) => {
+    console.log(sNum + ": Deleting old users...");
+    try {
+        // knex("users")
+        //     .where("snum", "<", sNum)
+        //     .del();
+        knex.raw('delete from users where snum <= ' + sNum + ";")
+            .catch(err => {
+                console.error(err);
+                throw err;
+            });
```

Comment:
why use raw?

Reply:
You can't use comparative where statements in knex